support/cas-server-support-oauth-core-api/src/main/java/org/apereo/cas/support/oauth/web/endpoints/OAuth20RevocationEndpointController.java

```diff
@@ -0,0 +1,191 @@
+            return OAuth20Utils.writeError(response, OAuth20Constants.INVALID_REQUEST);
+        }
+
+        val manager = new ProfileManager<CommonProfile>(context, context.getSessionStore());
+        val clientId = OAuth20Utils.getClientIdAndClientSecret(context).getLeft();
+        val registeredService = getRegisteredServiceByClientId(clientId);
+
+        if (isServiceNeedAuthentication(registeredService)) {
+            if (manager.get(true).isEmpty()) {
+                LOGGER.error("Service [{}] requests authentication", clientId);
+                return OAuth20Utils.writeError(response, OAuth20Constants.ACCESS_DENIED);
+            }
+        }
+        val token = context.getRequestParameter(OAuth20Constants.TOKEN)
+            .map(String::valueOf).orElse(StringUtils.EMPTY);
+
+        if (isAsupportedTokenType(token)) {
+            return generateRevocationResponse(token, clientId, response);
+        }
+        return OAuth20Utils.writeError(response, OAuth20Constants.INVALID_REQUEST);
+    }
+
+    /**
+     * Generate revocation token response.
+     *
+     * @param token the token to revoke
+     * @param client the client who requests the revocation
+     * @param response the response
+     * @return the model and view
+     */
+    private ModelAndView generateRevocationResponse(final String token,
+                                                    final String client,
+                                                    final HttpServletResponse response) {
+
+        if (token.startsWith(OAuth20RefreshToken.PREFIX)) {
+            val registryToken = getOAuthConfigurationContext().getTicketRegistry().getTicket(token, OAuth20RefreshToken.class);
+
+            if (registryToken == null) {
+                LOGGER.error("Provided token [{}] has not been found in the ticket registry", token);
+            } else {
+                if (!StringUtils.equals(client, registryToken.getClientId())) {
+                    LOGGER.error("Provided token [{}] is not related with the service [{}]", token, client);
+                    return OAuth20Utils.writeError(response, OAuth20Constants.INVALID_REQUEST);
+                }
+                revokeToken(registryToken);
+            }
+        } else {
+            val registryToken = getOAuthConfigurationContext().getTicketRegistry().getTicket(token, OAuth20AccessToken.class);
+
+            if (registryToken == null) {
+                LOGGER.error("Provided token [{}] has not been found in the ticket registry", token);
+            } else {
+                if (!StringUtils.equals(client, registryToken.getClientId())) {
+                    LOGGER.error("Provided token [{}] is not related with the service [{}]", token, client);
+                    return OAuth20Utils.writeError(response, OAuth20Constants.INVALID_REQUEST);
+                }
+                revokeToken(registryToken);
```

Comment:
All of this is duplicate code. You should simply fetch OAuth20Token from the registry as the type, and then attempt to cast where necessary to the right type.

Reply:
Sorry about that...
I'm still beginning with Java and still not comfortable with some of it's concepts... Hopefully work on CAS helped me a lot and still does ! :)

Reply:
Done.